### PR TITLE
[MLIR][DebugInfo] Set debug info format in MLIR->IR translation

### PR DESCRIPTION
**Pull request by SLTozer**
MLIR's LLVM dialect does not internally support debug records, only converting to/from debug intrinsics. To smooth the transition from intrinsics to records, there is a step prior to IR->MLIR translation that switches the IR module to intrinsic-form; this patch adds the equivalent conversion to record-form at MLIR->IR translation.

This is a partial reapply of https://github.com/llvm/llvm-project/pull/95098 which can be landed once the flang frontend has been updated by https://github.com/llvm/llvm-project/pull/95306. This is the counterpart to the earlier patch https://github.com/llvm/llvm-project/pull/89735 which handled the IR->MLIR conversion.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ConvertToLLVMIR.cpp [hunk 1]
```diff
@@ -31,6 +34,13 @@ void registerToLLVMIRTranslation() {
         if (!llvmModule)
           return failure();
 
+        // When printing LLVM IR, we should convert the module to the debug info
+        // format that LLVM expects us to print.
+        // See https://llvm.org/docs/RemoveDIsDebugInfo.html
+        llvm::ScopedDbgInfoFormatSetter FormatSetter(*llvmModule,
```
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ConvertToLLVMIR.cpp:40]:
```suggestion
        llvm::ScopedDbgInfoFormatSetter formatSetter(*llvmModule,
```
ultra nit: variables should start with lowercase in mlir.

File: mlir/test/lib/Dialect/Test/TestToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -122,6 +125,13 @@ void registerTestToLLVMIR() {
         if (!llvmModule)
           return failure();
 
+        // When printing LLVM IR, we should convert the module to the debug info
+        // format that LLVM expects us to print.
+        // See https://llvm.org/docs/RemoveDIsDebugInfo.html
+        llvm::ScopedDbgInfoFormatSetter FormatSetter(*llvmModule,
```
Comment (gysit) [inline mlir/test/lib/Dialect/Test/TestToLLVMIRTranslation.cpp:131]:
```suggestion
        llvm::ScopedDbgInfoFormatSetter formatSetter(*llvmModule,
```

**Review by gysit — APPROVED** [review]
Thanks LGTM modulo ultra nits!

**Review by walter-erquinigo — COMMENTED** [review]
nice!

**Pull request merged**